### PR TITLE
:book: Adding OLM V1 architecture document

### PR DESCRIPTION
**Pull request by LalatenduMohanty**
The documentation contains description of the components and a diagram. Removed the components.md file as the architecture doc contains the same information

Fixes : [#1124](https://github.com/operator-framework/operator-controller/issues/1124)

# Description

## Reviewer Checklist

- [ ] API Go Documentation
- [ ] Tests: Unit Tests (and E2E Tests, if appropriate)
- [ ] Comprehensive Commit Messages
- [ ] Links to related GitHub Issue(s)

### CODE REVIEW

File: docs/drafts/architecture.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+
+## OLM V1 Architecture
+
+This document talks about the V1 architecture. OLM V1 consists of many components, however majority of the code is divided in to two components
```
Comment (perdasilva) [inline docs/drafts/architecture.md:4]:
```suggestion
This document describes the OLM v1 architecture. OLM v1 consists of two main components:
```

File: docs/drafts/architecture.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+* Resolver in operator-controller helps the extension controller to filter the bundle reference after applying the user restrictions (e.g. name, priority etc) and returns the bundle reference to the extension controller.
+* Cluster extension controller reaches out to the registry to download the bundle container images, saves it to the bundle cache,  unpacks it and applies the bundle manifests to the cluster.  It is also Responsible for figuring out which bundle to upgrade.
+* Bundle cache returns the cache for the bundle. If a cache does not already exist, a new one will be created.
+
+### Catalogd:
+
+Catalogd unpacks [file-based catalog (FBC)](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#docs) content that is packaged and shipped in container images, for consumption by clients on-clusters (unpacking from other sources, like git repos, OCI artifacts etc, are in the roadmap for catalogD). As component of the Operator Lifecycle Manager (OLM) v1 microservices architecture, catalogD hosts metadata for Kubernetes extensions packaged by the authors of the extensions, as a result helping customers discover installable content.
```
Comment (perdasilva) [inline docs/drafts/architecture.md:22]: I'd move `(unpacking from other sources, like git repos, OCI artifacts etc, are in the roadmap for catalogD)` down to the end of the section. It's a great call out, I just think it's misplaced here.

Comment (perdasilva) [inline docs/drafts/architecture.md:22]:
```suggestion
Catalogd unpacks [file-based catalog (FBC)](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#docs) content that is packaged and shipped in container images, for consumption by clients on-clusters (unpacking from other sources, like git repos, OCI artifacts etc, are in the roadmap for catalogD). It serves the extension metadata, provided by the extension authors, found in the FBC, making it possible for on-cluster clients to discover installable content.
```

File: docs/drafts/architecture.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+* Catalogd can be broken down in to three sub-components i.e. ClusterCatalog controller, catalogd http server, catalogd content cache.
+* Catalog controller  is responsible for pulling FBC based catalog images from registry and unpacking them into the catalog content cache. It is also responsible for reconciling the latest changes in the cluster catalog.
+* Catalogd http server is responsible for serving catalog information to clients e.g. cluster extension controller.
+* Catalogd content cache is maintained by the catalog controller and used by the catalogd http server to answer queries from clients.
+
+
+
+### Diagram
```
Comment (perdasilva) [inline docs/drafts/architecture.md:31]: I'd suggest moving the diagram to the top.

**Review by perdasilva — COMMENTED** [review]
In general, I think it's really good. I've made a couple of suggestions and suggested edits. The suggested edits can also probably be improved. By hope they convey what I was thinking. I think we should also call out the CRs each component reconciles (with links to the API description).

File: docs/drafts/architecture.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+### Operator-controller:
+
+operator-controller is the central component of OLM v1, that consumes all of the components below to extend Kubernetes to allows users to install, and manage the lifecycle of other extensions
+
+* Operator-controller queries the catalogd (catalogd HTTP Server) to get catalog information.  Once received the catalog information is saved  to catalog-cache. The cache will be updated automatically if a Catalog is noticed to have a different resolved image reference.
+* Resolver in operator-controller helps the extension controller to filter the bundle reference after applying the user restrictions (e.g. name, priority etc) and returns the bundle reference to the extension controller.
+* Cluster extension controller reaches out to the registry to download the bundle container images, saves it to the bundle cache,  unpacks it and applies the bundle manifests to the cluster.  It is also Responsible for figuring out which bundle to upgrade.
+* Bundle cache returns the cache for the bundle. If a cache does not already exist, a new one will be created.
```
Comment (everettraven) [inline docs/drafts/architecture.md:18]: Feels odd that we are talking about the components by name before a reader sees the diagram

Comment (LalatenduMohanty) [reply docs/drafts/architecture.md]: Fixed it in https://github.com/operator-framework/operator-controller/pull/1301/commits/67c1540f1ed18d59b3dbc31f0fb59ff2a8a95eb4

**Comment by LalatenduMohanty** [comment]
> I think we should also call out the CRs each component reconciles (with links to the API description).

File: docs/drafts/architecture.md (hunk 2 above)
Comment (LalatenduMohanty) [reply docs/drafts/architecture.md]: This is resolved

**Review by everettraven — APPROVED** [review]

**Comment by perdasilva** [comment]
> > I think we should also call out the CRs each component reconciles (with links to the API description).
> 
> @perdasilva I do not want to go to that low level in this document as it would impact the readability of the doc. We can have another doc (as follow up) which can talk about this. So that folks want to go little low level can move to those docs after going through the architecture doc.

I get that. I just would have expected the API the users use to not really be a low level thing. It's the front door XD

**Comment by LalatenduMohanty** [comment]
@perdasilva We already have https://github.com/operator-framework/operator-controller/pull/1230 and https://github.com/operator-framework/operator-controller/pull/1240. 

So I am not entirely sure what you are looking for.  If you can create an issue with more details/context I can take a look to fix it.

**Pull request merged**